documents/model.go

```diff
@@ -754,3 +755,154 @@ func isNFTInRole(role *coredocumentpb.Role, registry common.Address, tokenID []b
+	tm = append(tm, grantee...)
+	tm = append(tm, roleID)
+	tm = append(tm, docID...)
+
+	// fetch key pair from identity
+	privateKey := id.Keys[identity.KeyPurposeEthMsgAuth].PrivateKey
+	pubKey := id.Keys[identity.KeyPurposeEthMsgAuth].PublicKey
+
+	// sign the assembled access token message, return signature
+	sig, err := crypto.SignMessage(privateKey, tm, "CurveSecp256K1", true)
+	if err != nil {
+		return nil, err
+	}
+	// assemble the access token, appending the signature and public keys
+	at := new(coredocumentpb.AccessToken)
+	at.Identifier = tokenIdentifier
+	at.Granter = granterID
+	at.Grantee = grantee
+	at.RoleIdentifier = []byte{roleID}
+	at.DocumentIdentifier = docID
+	at.Signature = sig
+	at.Key = pubKey
+
+	return at, nil
+}
+
+// AddAccessTokenToReadRules adds the AccessToken(s) to the read rules of the document
+func (m *CoreDocumentModel) AddAccessTokenToReadRules(id identity.IDConfig, payload documentpb.AccessTokenParams) error {
+	at, err := assembleAccessToken(m, id, payload)
+	if err != nil {
+		return errors.New("failed to construct AT: %v", err)
+	}
+
+	role := new(coredocumentpb.Role)
+	rk, err := utils.ConvertIntToByte32(len(m.Document.Roles))
+	if err != nil {
+		return err
+	}
+	role.RoleKey = rk[:]
+	role.AccessTokens = append(role.AccessTokens, at)
+	m.addNewRule(role, coredocumentpb.Action_ACTION_READ)
+	err = m.setCoreDocumentSalts()
+	if err != nil {
+		return errors.New("failed to generate CoreDocumentSalts")
+	}
+	return nil
+}
+
+// ATOwnerCanRead checks if the owner of the AT can read the document
+func (m *CoreDocumentModel) ATOwnerCanRead(tokenReq p2ppb.AccessTokenRequest , account identity.CentID) error {
+	// check if the access token is present in read rules
+	token, err := m.findAT(coredocumentpb.Action_ACTION_READ, tokenReq.AccessTokenId, func(role *coredocumentpb.Role, tokenID []byte) (*coredocumentpb.AccessToken, error){
+		return isATInRole(role, tokenReq.AccessTokenId)
+	})
+	if err != nil {
+		return err
+	}
+
+	granterID, err := identity.ToCentID(token.Granter)
+	if err != nil {
+		return err
+	}
+	// check that the token granter has the correct access type for the document
+	validGranter := m.AccountCanRead(granterID)
+	if !validGranter {
+		return errors.New("invalid access token granter")
+	}
+	// TODO:  check that the granter is the owner of the public key on the access token
+	
+	// check that the account requesting access is the grantee of the access token
+	if !bytes.Equal(account[:], token.Grantee) {
+		return errors.New("access token grantee is not the same as the account requesting access")
+	}
+	// validate the access token
+	err = validateAT(token.Key, token)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// validateAT validates that given access token against its signature
+// ethereum specific variation is currently false
+
+func validateAT (publicKey []byte, token *coredocumentpb.AccessToken) error {
+	tm := append(token.Identifier, token.Granter...)
```

Comment:
I think it would be a good idea to create a method that generates the signaturePayload and use that message both when creating the token and when validating the token instead of duplicating it here.